tailcfg/tailcfg.go

```diff
@@ -574,9 +584,47 @@ func (req *RegisterRequest) Clone() *RegisterRequest {
 		tok := *res.Auth.Oauth2Token
 		res.Auth.Oauth2Token = &tok
 	}
+	if res.DeviceCert != nil {
+		res.DeviceCert = make([]byte, len(req.DeviceCert))
+		copy(res.DeviceCert, req.DeviceCert)
+	}
+	if res.Signature != nil {
+		res.Signature = make([]byte, len(req.Signature))
+		copy(res.Signature, req.Signature)
+	}
 	return res
 }
 
+// WritePlaintextForSigning writes the signed fields (along with the server and machine
+// public keys) to w needed to sign or verify a RegsterRequest.
+func (req *RegisterRequest) WritePlaintextForSigning(
+	w io.Writer, serverURL string, serverPubKey, machinePubKey wgkey.Key) error {
```

Comment:
you probably only want to do one Write to their Writer.

fmt.Fprintf will do that for you, which would also consolidate all your error checks down into one.

Reply:
OK.

Is fmt.Fprintf ok with writing out non-UTF8 binary stuff (e.g. cert, keys)? I suppose it doesn't actually matter if it hex encodes the resulting data or not (it's all just opaque data being fed into sha256 so it just needs to be consistent on both sides) but that probably involves extra allocations.

I could probably also skip the error checking because Write() implemented for hash says it will never return an error anyway.

Reply:
It doesn't care about UTF-8. a %s of binary will print binary.

Reply:
Ah I remember how I ended up doing it this way -- if I combine everything into a big write it'll require building a big temporary buffer for the whole message, whereas separate writes could avoid making a copy. SHA256 does need data in a certain chunk size though so it'll probably end up copying to a fixed size internal buffer anyway. I suppose it probably doesn't make a big difference either way.

Reply:
`fmt` has an internal buffer pool that it re-uses between writes.

If you really want to do a bunch of little writes, you shouldn't take an `io.Writer`. You should instead take a `*bufio.Writer`, or document that you do a bunch of little writes. Otherwise you risk somebody far away being surprised by a bunch of network writes or disk writes.

But is this only used for hashing? Maybe take a `hash.Hash` then instead which shows intent a bit more, about what you expect to pass to it, even if it's an unnecessarily capable interface?

Reply:
Alright, switched it to do both: use fmt.Fprint (Fprintf with a format string of %s%s%s%s%s... seems annoying, and the extra spaces provide a delimiter which may be helpful and are at worst harmless), and take a `hash.Hash`. I'd slightly prefer `crypto.Hash` because then I could directly pass `crypto.SHA256` but `crypto` seems like a heavier dependency than plain `hash`.

I think the result looks more appropriate compared to the rest of tailcfg too.

Reply:
This code looks good but I still don’t like it being in tailcfg. Any oss package is easily reachable from the server. You can even make a new one if you like.

Reply:
ok moved out of tailcfg to controlclient